### PR TITLE
Improve AzureCoreEventSource performance as well as its usage

### PR DESCRIPTION
**Pull request by AlexanderSher**
According to this [recommendation](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs#L1347-L1362), performance can be increased with additional helper methods.

Also, as @KrzysztofCwalina pointed out, we don't need to calculate arguments for the events that which level is disabled

### CODE REVIEW

File: sdk/core/Azure.Core/src/Diagnostics/AzureCoreEventSource.cs [hunk 1]
```diff
@@ -37,6 +37,12 @@ internal sealed class AzureCoreEventSource : AzureEventSource
 
         public static AzureCoreEventSource Singleton { get; } = new AzureCoreEventSource();
 
+        [NonEvent]
+        public bool IsEnabledInformational() => IsEnabled(EventLevel.Informational, EventKeywords.None);
```
Comment (lmolkova) [inline sdk/core/Azure.Core/src/Diagnostics/AzureCoreEventSource.cs:41]: nit: if we find these methods useful in core, they can be useful in other sources and could be moved to the base `AzureEventSource`

File: sdk/core/Azure.Core/src/Pipeline/HttpPipelineBuilder.cs [hunk 2]
```diff
@@ -183,7 +183,7 @@ void AddNonNullPolicies(HttpPipelinePolicy[] policiesToAdd)
             {
                 if (buildOptions.ClientOptions.IsCustomTransportSet)
                 {
-                    if (AzureCoreEventSource.Singleton.IsEnabled())
+                    if (AzureCoreEventSource.Singleton.IsEnabledInformational())
```
Comment (annelo-msft) [inline sdk/core/Azure.Core/src/Pipeline/HttpPipelineBuilder.cs:186]: Curious: how do we know that this doesn't change the behavior from what it was previously?  Was there a guarantee that this was only logged at the informational level?

Comment (AlexanderSher) [reply sdk/core/Azure.Core/src/Pipeline/HttpPipelineBuilder.cs:186]: Each event method is marked with `EventAttribute` which specifies event level. So these added checks match those levels.

[the remark this answers is not in the document]
Comment (AlexanderSher) [reply sdk/core/Azure.Core/src/Pipeline/HttpPipelineBuilder.cs:186]: Agreed. However, we want the check to be done before all the reformatting of headers is happening. But I can move that logic inside `AzureCoreEventSource` as well.

[the remark this answers is not in the document]
Comment (jsquire) [reply sdk/core/Azure.Core/src/Pipeline/HttpPipelineBuilder.cs:186]: I'm not opposed to having an external check here as well, if we can't reasonably move this into the event source.    If we decide to keep that here, we should consider a comment detailing why we've done it and noting that it is redundant to the check in the event source.

File: sdk/core/Azure.Core/src/Diagnostics/AzureCoreEventSource.cs (hunk 1 above)
Comment (AlexanderSher) [reply sdk/core/Azure.Core/src/Diagnostics/AzureCoreEventSource.cs]: Moved `WriteEvent` helpers to the `AzureEventSource`

File: sdk/core/Azure.Core/src/Pipeline/HttpPipelineBuilder.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (AlexanderSher) [reply sdk/core/Azure.Core/src/Pipeline/HttpPipelineBuilder.cs]: I have added non-event helper methods to the `AzureCoreEventSource` that wraps. They don't replace event methods, but at least they are in the same type and are right next to the called events.

**Comment by AlexanderSher** [comment]
I have added a perf test. However, it shows that [said recommendation](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs#L1347-L1362) is incorrect, at least in our scenario. On two different machines, after multiple runs, results are pretty much similar

|                              Method |      Runtime |      Mean | Ratio | RatioSD |   Gen0 | Allocated | Alloc Ratio |
|------------------------------------ |------------- |----------:|------:|--------:|-------:|----------:|------------:|
|                  Old implementation |     .NET 6.0 |  5.937 us |  1.00 |    0.00 | 0.1831 |   6.38 KB |        1.00 |
|   New implementation + reformatting |     .NET 6.0 |  6.137 us |  1.03 |    0.02 | 0.1984 |   6.83 KB |        1.07 |
| New implementation, no reformatting |     .NET 6.0 |  4.210 us |  0.71 |    0.02 | 0.1450 |   5.07 KB |        0.80 |
|                                     |              |           |       |         |        |           |             |
|                  Old implementation | .NETFX 4.6.1 | 12.118 us |  1.00 |    0.00 | 4.2419 |  17.39 KB |        1.00 |
|   New implementation + reformatting | .NETFX 4.6.1 | 22.480 us |  1.86 |    0.04 | 4.4556 |  18.36 KB |        1.06 |
| New implementation, no reformatting | .NETFX 4.6.1 | 17.606 us |  1.45 |    0.03 | 3.9368 |  16.13 KB |        0.93 |

Unless someone can find an error in the test, I'm going to remove custom `WriteEvent` methods

**Comment by KrzysztofCwalina** [comment]
> Unless someone can find an error in the test, I'm going to remove custom WriteEvent methods

The benchmarks seem to have LogAlways level enabled. I was thinking that you would see perf savings when some events are enabled but not the request details events, e.g. when only errors are logged/observed.

File: sdk/core/Azure.Core/src/Shared/AzureEventSource.cs [hunk 3]
```diff
@@ -85,5 +85,243 @@ private static string DeduplicateName(string eventSourceName)
 
             return eventSourceName;
         }
+
+        [NonEvent]
+        protected unsafe void WriteEvent(int eventId, string? arg0, double arg1)
```
[the remark this answers is not in the document]
Comment (AlexanderSher) [reply sdk/core/Azure.Core/src/Shared/AzureEventSource.cs]: See benchmarks below. Methods are removed.

**Review by christothes — APPROVED** [review]

**Pull request merged**